### ISSUE TITLE
`admin_authenticator` is poorly documented, leading to insecure implementations

### ISSUE THREAD

**Issue by RobinDaugherty** [doorkeeper-gem/doorkeeper#1097]
In the [example initializer](https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/generators/doorkeeper/templates/initializer.rb#L5), the naming, comments, and examples provided for `resource_owner_authenticator` and `admin_authenticator` make it seem like they function similarly.

```rb
  # This block will be called to check whether the resource owner is authenticated or not.
  resource_owner_authenticator do
    # Put your resource owner authentication logic here.
    # Example implementation:
    #   User.find_by_id(session[:user_id]) || redirect_to(new_user_session_url)
```

`resource_owner_authenticator` returns the resource owner object, and if it returns `nil`, there will be no user signed in.

```rb
  # If you want to restrict access to the web interface for adding oauth authorized applications,
  # you need to declare the block below.
  #
  # admin_authenticator do
  #   # Put your admin authentication logic here.
  #   # Example implementation:
  #   Admin.find_by_id(session[:admin_id]) || redirect_to(new_admin_session_url)
```

However, `admin_authenticator` is used only as a `before_action` so it must raise an exception or call `redirect`, `render`, etc. if the permission should be denied. If it returns nil, the user is granted access.

I've seen a number of bad implementations of this in production applications. The most common implementation is something like this:

```rb
  admin_authenticator do
    user = Admin.find_by_id(session[:admin_id])
    (user || redirect_to(new_admin_session_url)) && user.may_access_admin?
  end
```

There seems to be no recommended application integration test for these endpoints, so most unexperienced developers would deploy this and be ignorant of the fact that the admin interface is open to all users. This is compounded by the fact that many installations wouldn't actually link to this interface, so it's not likely to be caught.

Another bad implementation is something like this:

```rb
  admin_authenticator do
    user = Admin.find_by_id(session[:admin_id])
    redirect_to(new_admin_session_url) unless user && user.may_access_admin?
  end
```

This can be bad in two ways. First, the correct behavior here would be to return a Forbidden status code or something similar. A Redirect HTTP status is technically wrong, but it can also lead to other problems. For example, if the Sign In page redirects back if a user is already authenticated, they will end up in an infinite redirection loop.

Additionally, the fact that there is no `admin_authenticator` implementation by default means that most applications are going to be implemented with all users able to create and edit OAuth applications, which does not seem like a good default.

In short, these incorrect examples that leave out important information and lead beginners to implement bad practices or deploy applications that are insecure by default.

**Comment by nbulaj** [comment]
Hi @RobinDaugherty . I totally agree with the necessity of more robust documentation. Would you like to submit a PR ?

What about the default behaviour - I was thinking that it doesn't allow to manage applications if block returned nil (so nothing insecure there), but I will check it tomorrow.

**Comment by nbulaj** [comment]
Hi @RobinDaugherty . I've checked default behavior of `admin_authenticator` block and you are right, it is unsafe by default. So I've added a default block to this option with forbidden method in order to force developers to define their own authenticators.

You can check out 07afb4d67b36b0f60c9f8ce25d08b8a7769e900b . Do you still want to improve the docs?

**Comment by RobinDaugherty** [comment]
Glad to see that change! Here's my `admin_authenticator` if you're looking for a real-world example:

```rb
  admin_authenticator do
    if current_user
      head :forbidden unless current_user.is_sysadmin?
    else
      redirect_to(sign_in_url)
    end
  end
```

**Comment by nbulaj** [comment]
Added in b55c0adda373faae2661e5772705869cbc895b12, thanks!

**Comment by kreintjes** [comment]
The part in the documentation also isn't really clear about this: https://doorkeeper.gitbook.io/guides/ruby-on-rails/configuration

I think it would be wise to explicitly state the authenticator block shouldn't return a boolean, but redirect/raise if the user isn't authenticated. We almost did this wrong when following this documentation.

**Comment by nbulaj** [comment]
Hi @kreintjes . I'm for any docs improvements! Would you luke to propose a PR?

**Comment by RobinDaugherty** [comment]
Please reopen this issue. The documentation continues to tell people that the Doorkeeper admin area is secure by default, and that the `admin_authenticator` block "follows the same rules as the resource_owner_authenticator block".

This is completely insecure. All application developers that use Doorkeeper currently think that their `admin_authenticator` block is securing the oauth admin area, when the mere addition of that block **makes the admin area accessible no matter what you return**.

**Comment by nbulaj** [comment]
Could you please clarify what do you mean?

By default (without configuration) this block returns Forbidden
https://github.com/doorkeeper-gem/doorkeeper/blob/e048bedb7813bb273e05554a66cc1c0411638d44/lib/doorkeeper/config.rb#L244

You should return the error by yourself in that block, not the user instance or true/false

**Comment by RobinDaugherty** [comment]
Sorry, to clarify I meant "by default when they declare `admin_authenticator`". In any application where the provide the Doorkeeper admin interface, they would likely not want all users to have access to the admin interface. So this block will be declared in a lot of cases. And the moment this block is declared, no matter what it returns, even if it is empty, access is granted to all users.

The configuration template and the documentation are both misleading or unclear about this.

Documentation says that it "follows the same rules as the resource_owner_block", when it does not. That block uses the return value of the block, while `admin_authenticator` does not. Returning nil from the `resource_owner` block will result in no access, while doing the same from `admin_authenticator` grants all access.

The example code in the template demonstrates the right thing (calling `head` or `redirect_to`), it doesn't explain that this is necessary. The call pattern of the block does unexplained, and the documentation is actually misleading in a way that causes an insecure implementation.
